typescript/src/client/client.ts

```diff
@@ -0,0 +1,261 @@
+import { v4 as uuuid } from "uuid";
+import { ErrorModel } from "../models/errors.js";
+import {
+  Agent,
+  AgentName,
+  AwaitResume,
+  Event,
+  Run,
+  RunId,
+  SessionId,
+} from "../models/models.js";
+import {
+  AgentsListResponse,
+  AgentsReadResponse,
+  PingResponse,
+  RunCreateRequest,
+  RunCreateResponse,
+  RunEventsListResponse,
+  RunReadResponse,
+  RunResumeRequest,
+  RunResumeResponse,
+} from "../models/schemas.js";
+import { ACPError, BaseError, FetchError, HTTPError } from "./errors.js";
+import { createEventSource, EventSource } from "./sse.js";
+import { Input } from "./types.js";
+import { inputToMessages } from "./utils.js";
+
+type FetchLike = typeof fetch;
+
+interface ClientInit {
+  baseUrl?: string;
+  /**
+   * Optional fetch implementation to use. Defaults to `globalThis.fetch`.
+   * Can also be used for advanced use cases like mocking, proxying, custom certs etc.
+   */
+  fetch?: FetchLike;
+  sessionId?: string;
+}
+
+export class Client {
+  #baseUrl: string;
+  #fetch: FetchLike;
+  #sessionId?: SessionId;
+
+  constructor(init?: ClientInit) {
+    this.#fetch = init?.fetch ?? globalThis.fetch;
+    this.#baseUrl = normalizeBaseUrl(init?.baseUrl ?? "");
+    this.#sessionId = init?.sessionId;
+  }
+
+  get sessionId() {
+    return this.#sessionId;
+  }
+
+  withSession(sessionId: SessionId = uuuid()) {
+    return new Client({
+      fetch: this.#fetch,
+      baseUrl: this.#baseUrl,
+      sessionId,
+    });
+  }
```

Comment:
The problem I see with this interface is that we cannot automatically execute some code after the session is no longer needed. This is necessary, for example, with the opentelemetry instrumentation which you should include here to tie multiple runs together into one trace.

I'm afraid we must go the `await withSession(async (session) => {})` route although I too dislike the pattern.